alf/utils/video_recorder.py

```diff
@@ -179,23 +221,199 @@ def capture_frame(self, pred_info=None):
+        to be shown in the video frames.
+
+        Args:
+            info_func (None|callable): a callable for calculating some customized
+                information (e.g. predicted future reward) based on the observation
+                at each time step and action sequences from the current time step
+                to the next ``future_steps`` steps (if available). It is called
+                as ``pred_info=info_func(current_observation, action_sequences)``.
+                Currently only support displaying scalar predictive information
+                returned from info_func.
+            encode_all (bool): whether to encode all the steps in the episode
+                buffer (i.e. the list of observations/actions/rewards).
+                - If False, only encode one step. In this case, if
+                    ``future_steps`` is smaller than the length of the episode
+                    buffer, one step of defer encoding will be conducted.
+                - If True, encode all the steps in episode_buffer. In this case,
+                    the actual ``future_steps`` is upper-bounded by the
+                    length of the episode buffer - 1.
+        """
+        # [episode_buffer_length, reward_dim]
+        rewards = torch.cat(self._reward_buffer, dim=0)
+        episode_buffer_length = rewards.shape[0]
+
+        if not encode_all and self._future_steps >= episode_buffer_length:
+            # not enough future date for defer encoding
+            return
+
+        if rewards.ndim > 1:
+            # slice the multi-dimensional rewards
+            # assume the first dimension is the overall reward
+            rewards = rewards[..., 0]
+
+        actions = torch.cat(self._action_buffer, dim=0)
+
+        num_steps = self._future_steps + 1
+
+        reward_curve_set = []
+        action_curve_set = []
+        predictive_curve_set = []
+
+        encoding_steps = episode_buffer_length if encode_all else 1
+        for t in range(encoding_steps):
+            H = min(num_steps,
+                    episode_buffer_length - t)  # total display steps
+            if H > 0:
+                t_actions = actions[t:t + H]
+                t_rewards = rewards[t:t + H]
+
+                if info_func is not None:
+                    predictions = info_func(self._observation_buffer[t],
+                                            t_actions)
+                    assert predictions.ndim == 1 or predictions.shape[1] == 1, \
+                        "only support displaying scalar predictive information"
+                    predictions = predictions.view(-1).detach().cpu().numpy()
+
+                    pred_curve = self._plot_value_curve(
+                        "prediction", [predictions],
+                        legends=["Prediction"],
+                        fig_size=6,
+                        height=300,
+                        width=300,
+                        linewidth=5)
```

Comment:
Maybe these common constant argument numbers should be variables for reuse.

Reply:
Done.